File: barbearia-app/src/app/services/servico/servico.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Servico } from '../../model/servicos/servico.model';

@Injectable({
  providedIn: 'root'
})
export class ServicoService {

  private apiUrl = 'http://localhost:8080/api/servicos';

  constructor(private http: HttpClient) { }

  save(servico: Servico): Observable<Servico> {
    return this.http.post<Servico>(this.apiUrl, servico);
  }

  findAll(): Observable<Servico[]> {
    return this.http.get<Servico[]>(this.apiUrl);
  }

  findById(id: number): Observable<Servico> {
    return this.http.get<Servico>(`${this.apiUrl}/${id}`);
  }

  update(id: number, servico: Servico): Observable<Servico> {
    return this.http.put<Servico>(`${this.apiUrl}/${id}`, servico);
  }

  deleteById(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  buscarPorNome(nome: string): Observable<Servico[]> {
    return this.http.get<Servico[]>(`${this.apiUrl}/nome/${nome}`);
  }

  buscarPorPrecoAte(preco: number): Observable<Servico[]> {
    return this.http.get<Servico[]>(`${this.apiUrl}/preco/${preco}`);
  }

  buscarQtdServico(): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/quantidade`);
  }
}

File: Barbearia4Periodo-master/Barbearia4Periodo-master/src/main/java/app/service/ServicoService.java
package app.service;

import app.entity.Servico;
import app.repository.ServicoRepository;
import jakarta.persistence.EntityNotFoundException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ServicoService {

    private final ServicoRepository servicoRepository;

    @Autowired
    public ServicoService(ServicoRepository servicoRepository) {
        this.servicoRepository = servicoRepository;
    }

    // Create
    public Servico save(Servico servico) {
        try {
            return servicoRepository.save(servico);
        } catch (Exception e) {
            e.printStackTrace();  // Log do erro completo
            throw new RuntimeException("Error saving Servico", e);  // Lançar uma exceção com mensagem
        }
    }

    // Read
    public List<Servico> findAll() {
        return servicoRepository.findAll();
    }

    public Servico findById(Long id) {
        return servicoRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Servico not found with id: " + id));
    }

    // Update
    public Servico update(Long id, Servico servicoDetails) {
        Servico servico = findById(id);
        servico.setNome(servicoDetails.getNome());
        servico.setDescricao(servicoDetails.getDescricao());
        servico.setPreco(servicoDetails.getPreco());
        servico.setDuracaoMinutos(servicoDetails.getDuracaoMinutos());
        return servicoRepository.save(servico);
    }
   
    public List<Servico> buscarPorNome(String nome) {
        return servicoRepository.findByNomeIgnoreCaseContaining(nome);
    }

    public List<Servico> buscarPorPrecoAte(java.math.BigDecimal preco) {
        return servicoRepository.findByPrecoLessThanEqual(preco);
    }
    
    // Delete
    public void deleteById(Long id) {
        if (!servicoRepository.existsById(id)) {
            throw new EntityNotFoundException("Servico not found with id: " + id);
        }
        servicoRepository.deleteById(id);
    }
    public Long buscarQtdServico(){
        return servicoRepository.buscarQtdServico();
    }


}
